### PR TITLE
Basic action wrapping for sails

### PR DESCRIPTION
**Pull request by achandras**
This should create a span for each action invoked by the router or actions called directly.

### CODE REVIEW

File: src/plugins/sails.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+const Tags = require('opentracing').Tags
+
+function wrapAction (action, tracer, identity) {
+  return function actionWithTracing (req, res, next) {
+    const classicAction = typeof action === 'function'
+
+    const scope = tracer.scope()
+    const childOf = scope.active()
+    const span = tracer.startSpan(
```
Comment (rmfitzpatrick) [inline src/plugins/sails.js]: Where do these spans reside w/ those of express and http?

File: test/plugins/sails.spec.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        })
+      })
+
+      after(() => {
+        Sails.lower()
+
+        agent.close()
+      })
+
+      it('traces registered actions', done => {
```
Comment (rmfitzpatrick) [inline test/plugins/sails.spec.js]: Can you add an actions2 test?

**Comment by rmfitzpatrick** [comment]
Have you used this with something like https://github.com/balderdashy/seed?

File: docs/API.md [hunk 3]
```diff
@@ -446,6 +446,20 @@ mysql2
 | http.status_code | The HTTP status code of the response.                     |
 | http.headers.*   | A recorded HTTP header.                                   |
 
+#### sails
+
+Much of the Sails instrumentation is provided by other component instrumentations,
+such as the DNS and HTTP instrumentations, Express, and the various database drivers.
+However, as those instrumentations might cause traces to have a large number of
+superfluous spans, consider selectively enabling the instrumentations desired when
+using Sails.
```
Comment (rmfitzpatrick) [inline docs/API.md]: This is helpful information but could it be more clear to just state what this instrumentation traces with a lighter note of its relationship w/ Express?

File: test/plugins/sails.spec.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        Sails.registerAction(() => { return 'test response' }, 'testAction')
+
+        expect(Sails.getActions()['testaction']()).to.eq('test response')
```
Comment (rmfitzpatrick) [inline test/plugins/sails.spec.js]: though it's really unlikely to flake, I think the convention of adding the expect handlers for trace content before initiating the traced action is done for a reason.

**Comment by achandras** [comment]
I have not tried it with seed, just a default generated Sails app. I can try it with that

File: src/plugins/sails.js (hunk 1 above)
Comment (achandras) [reply src/plugins/sails.js]: I think these are handled as express middleware, so express spans are before and after it

File: src/plugins/sails.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        }
+      }
+    )
+
+    let retVal
+    try {
+      // if the action is not a classic req/res or a machine, calling it will throw
+      const fn = classicAction ? action : action[0]
+      retVal = fn.apply(this, arguments)
+    } catch (e) {
```
[the remark this answers is not in the document]
Comment (achandras) [reply src/plugins/sails.js]: Nope, fixed. Thanks

File: docs/API.md [hunk 6]
```diff
@@ -446,13 +446,16 @@ mysql2
 | http.status_code | The HTTP status code of the response.                     |
 | http.headers.*   | A recorded HTTP header.                                   |
 
-##### Configuration Options
+#### sails
```
Comment (rmfitzpatrick) [inline docs/API.md]: This removes the restify configuration options by mistake?  Also looks like their branding prefers capital S.

File: test/plugins/sails.spec.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      })
+
+      it('traces registered actions2 actions', done => {
+        function action2 () {
+          return 'action2 response'
+        }
+        // clumsy way of faking the object that machine-as-action returns
+        action2.IS_MACHINE_AS_ACTION = true
```
Comment (rmfitzpatrick) [inline test/plugins/sails.spec.js]: Not sure I understand this, or why registering something like `{fn: action2}` isn't adequate?  Definitely don't know much about sails though, so if this is an internal constant a link would be helpful.

File: docs/API.md (hunk 6 above)
Comment (achandras) [reply docs/API.md]: My mistake, I will add that back in. Thanks

Do we follow branding in any of the other section titles? It appears to be just the name of the package for the other ones, but if you prefer proper names instead, we should change all of them.

File: test/plugins/sails.spec.js (hunk 7 above)
Comment (achandras) [reply test/plugins/sails.spec.js]: By the time it reaches this method it is no longer the dictionary I think you are thinking of. It gets processed though `machine-as-action`, so this is the easiest way to produce that object without actually running it through that module. The name of member is meaningless.

I can link to the module code if that's useful, but I'm not sure what else would be helpful.

File: docs/API.md (hunk 6 above)
Comment (rmfitzpatrick) [reply docs/API.md]: fair point.  I'll update the others in my api doc changes.

File: test/plugins/sails.spec.js (hunk 7 above)
Comment (achandras) [reply test/plugins/sails.spec.js]: I think I understand now. This is my fault, I'll fix the poor implementation I've done here.

Comment (achandras) [reply test/plugins/sails.spec.js]: There should now be a more clear distinction between the regular and middleware registration paths for controllers and this path of registering action2 dictionaries programmatically will actually work

**Pull request merged**